File: cpp/3ac.cpp
#include <bits/stdc++.h>
using namespace std;

class ThreeAddressCode {
    int tempCount;
    vector<string> code;

public:
    ThreeAddressCode() { tempCount = 0; }

    string newTemp() {
        tempCount++;
        return "t" + to_string(tempCount);
    }

    string generate(string expr) {
        expr.erase(remove(expr.begin(), expr.end(), ' '), expr.end());
        return genExpr(expr);
    }

    void printCode() {
        for (auto &line : code)
            cout << line << endl;
    }

private:
    string genExpr(string expr) {
        if (expr.empty()) return "";

        bool alnum_only = all_of(expr.begin(), expr.end(), [](char c){ return isalnum(c); });
        if (alnum_only) return expr;

        if (expr.front() == '(' && expr.back() == ')')
            return genExpr(expr.substr(1, expr.size() - 2));

        for (char op : {'+', '-'}) {
            int level = 0;
            for (int i = expr.size() - 1; i >= 0; i--) {
                if (expr[i] == ')') level++;
                else if (expr[i] == '(') level--;
                else if (level == 0 && expr[i] == op) {
                    string left = genExpr(expr.substr(0, i));
                    string right = genExpr(expr.substr(i+1));
                    string temp = newTemp();
                    code.push_back(temp + " = " + left + " " + op + " " + right);
                    return temp;
                }
            }
        }

        for (char op : {'*', '/'}) {
            int level = 0;
            for (int i = expr.size() - 1; i >= 0; i--) {
                if (expr[i] == ')') level++;
                else if (expr[i] == '(') level--;
                else if (level == 0 && expr[i] == op) {
                    string left = genExpr(expr.substr(0, i));
                    string right = genExpr(expr.substr(i+1));
                    string temp = newTemp();
                    code.push_back(temp + " = " + left + " " + op + " " + right);
                    return temp;
                }
            }
        }

        return expr;
    }
};

int main() {
    string expr = "a + b * (c - d) / e";
    ThreeAddressCode tac;
    tac.generate(expr);
    tac.printCode();
    return 0;
}

File: 3ac.py
class ThreeAddressCodeGenerator:
    def __init__(self):
        self.temp_count = 0
        self.code = []

    def new_temp(self):
        self.temp_count += 1
        return f"t{self.temp_count}"

    def generate(self, expr):
        """Generate 3AC for the given arithmetic expression."""
        expr = expr.replace(" ", "")  
        return self._gen_expr(expr)

    def _gen_expr(self, expr):
        
        if expr.isalnum():
            return expr

        
        if expr.startswith("(") and expr.endswith(")"):
            return self._gen_expr(expr[1:-1])

        
        for op in ['+', '-']:
            level = 0
            for i in range(len(expr)-1, -1, -1):
                if expr[i] == ')':
                    level += 1
                elif expr[i] == '(':
                    level -= 1
                elif level == 0 and expr[i] == op:
                    left = self._gen_expr(expr[:i])
                    right = self._gen_expr(expr[i+1:])
                    temp = self.new_temp()
                    self.code.append(f"{temp} = {left} {op} {right}")
                    return temp

        for op in ['*', '/']:
            level = 0
            for i in range(len(expr)-1, -1, -1):
                if expr[i] == ')':
                    level += 1
                elif expr[i] == '(':
                    level -= 1
                elif level == 0 and expr[i] == op:
                    left = self._gen_expr(expr[:i])
                    right = self._gen_expr(expr[i+1:])
                    temp = self.new_temp()
                    self.code.append(f"{temp} = {left} {op} {right}")
                    return temp

        return expr  


expr = "a + b * (c - d) / e"
tac_gen = ThreeAddressCodeGenerator()
result = tac_gen.generate(expr)

print("Three-Address Code:")
for line in tac_gen.code:
    print(line)

